tensorboard/plugins/debugger/interactive_debugger_server_lib.py

```diff
@@ -0,0 +1,435 @@
+    """
+    self._run_key_to_original_graphs = dict()
+    self._run_key_to_debug_graphs = dict()
+    self._maybe_expanded_node_names = dict()
+
+    if breakpoints_func:
+      assert callable(breakpoints_func)
+      self._breakpoints_func = breakpoints_func
+
+  def add_graph(self, run_key, device_name, graph_def, debug=False):
+    """Add a GraphDef.
+
+    Args:
+      run_key: A key for the run, containing information about the feeds,
+        fetches, and targets.
+      device_name: The name of the device that the `GraphDef` is for.
+      graph_def: An instance of the `GraphDef` proto.
+      debug: Whether `graph_def` consists of the debug ops.
+    """
+    graph_dict = (self._run_key_to_debug_graphs if debug else
+                  self._run_key_to_original_graphs)
+    if not run_key in graph_dict:
+      graph_dict[run_key] = dict()  # Mapping device_name to GraphDef.
+    graph_dict[run_key][tf.compat.as_str(device_name)] = graph_def
+
+  def get_graphs(self, run_key, debug=False):
+    """Get the runtime graphs associated with a run key.
+
+    Args:
+      run_key: A Session.run kay.
+      debug: Whether the debugger-decoratedgraph is to be retrieved.
+
+    Returns:
+      A `dict` mapping device name to `GraphDef` protos.
+    """
+    graph_dict = (self._run_key_to_debug_graphs if debug else
+                  self._run_key_to_original_graphs)
+    return graph_dict.get(run_key, {})
+
+  def get_graph(self, run_key, device_name, debug=False):
+    """Get the runtime graphs associated with a run key and a device.
+
+    Args:
+      run_key: A Session.run kay.
+      device_name: Name of the device in question.
+      debug: Whether the debugger-decoratedgraph is to be retrieved.
+
+    Returns:
+      A `GraphDef`.
+    """
+    return self.get_graphs(run_key, debug=debug).get(device_name, None)
+
+  def get_breakpoints(self):
+    """Obtain all the currently activated breakpoints."""
+    return self._breakpoints_func()
+
+  def get_maybe_base_expanded_node_name(self, node_name, run_key, device_name):
+    """Obtain possibly base-expanded node name.
+
+    Base-expansion is the transformation of a node name which happens to be the
+    name scope of other nodes in the same graph. For example, if two nodes,
+    called 'a/b' and 'a/b/read' in a graph, the name of the first node will
+    be base-expanded to 'a/b/(b)'.
+
+    This method uses caching to avoid unnecessary recomputation.
+
+    Args:
+      node_name: Name of the node.
+      run_key: The run key to which the node belongs.
+      graph_def: GraphDef to which the node belongs.
+    """
+    device_name = tf.compat.as_str(device_name)
+    if (run_key in self._maybe_expanded_node_names and
+        device_name in self._maybe_expanded_node_names[run_key] and
+        node_name in self._maybe_expanded_node_names[run_key][device_name]):
+      return self._maybe_expanded_node_names[run_key][device_name][node_name]
```

Comment:
Have you considered `self._maybe_expanded_node_names[(run_key, device_name)][node_name]`?